corehq/apps/locations/forms.py

```diff
@@ -241,6 +246,12 @@ def clean_site_code(self):
     def clean_location_type(self):
         loc_type = self.cleaned_data['location_type']
 
+        try:
+            loc_type_obj = LocationType.objects.get(Q(code=loc_type)|Q(name=loc_type))
```

Comment:
Sorry you're right I got confused between locations and types.

> My preference for actions to take here are to leave this as-is,
immediately add those unique_togethers, and once we're no longer using this
in bulk upload, switch to pk references. How does that sound?

sounds good to me, thanks!

On Wed, Sep 14, 2016 at 9:44 PM, Ethan Soergel <notifications@github.com>
wrote:

> In corehq/apps/locations/forms.py
> <https://github.com/dimagi/commcare-hq/pull/13135#discussion_r78781651>:
>
> > @@ -241,6 +246,12 @@ def clean_site_code(self):
> >      def clean_location_type(self):
> >          loc_type = self.cleaned_data['location_type']
> >
> > +        try:
> > +            loc_type_obj = LocationType.objects.get(Q(code=loc_type)|Q(name=loc_type))
>
> You're thinking of Locations, it looks like we don't actually enforce
> *any* uniqueness constraints on LocationTypes...
> That seems like a problem. I just checked on prod, and it looks possible
> to add a unique_together on both ('domain', 'name') and ('domain', 'code').
> There's only one real domain that needs to be modified, and it looks like a
> mistake that there's an issue anyways.
>
> (LocationType.objects
>  .values('domain', 'code')
>  .order_by()
>  .annotate(count=Count('pk'))
>  .filter(count__gt=1))
>
> As for this immediate issue, once Sravan releases the new bulk upload
> feature, LocationForm will no longer be used in bulk upload, and we can
> switch this to the unambiguous location_type_pk.
>
> My preference for actions to take here are to leave this as-is,
> immediately add those unique_togethers, and once we're no longer using
> this in bulk upload, switch to pk references. How does that sound?
>
> Sheel and I were also just talking about how the CRUD workflow around
> Locations needs a lot of love, and I think we'll soon do a redesign and
> rewrite.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/13135/files/73de0cd594f7d0392681295d32cd4301b5c23317#r78781651>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAED-0Zl4fWoTQHqaIQXeGbp56mfZkjrks5qqB1OgaJpZM4J3Ptd>
> .
>